File: data-mesh-taxonomy-py/taxonomy/validation.py
"""Validation rules and typed exceptions, mirrored across API and Excel import."""
from __future__ import annotations

from .types import Level, TECHNICAL_OWNER_REQUIRED


class AppError(Exception):
    status = 500

    def __init__(self, message: str) -> None:
        super().__init__(message)
        self.message = message

    def payload(self) -> dict:
        return {"error": self.message}


class ValidationError(AppError):
    status = 400

    def __init__(self, fields: dict[str, str], message: str = "Validation failed") -> None:
        super().__init__(message)
        self.fields = fields

    def payload(self) -> dict:
        return {"error": self.message, "fields": self.fields}


class ConflictError(AppError):
    status = 409


class NotFoundError(AppError):
    status = 404


class UnauthorizedError(AppError):
    status = 401


class ForbiddenError(AppError):
    status = 403


def normalize_technical_owner(value: str | None) -> str | None:
    if value is None:
        return None
    trimmed = value.strip()
    return trimmed or None


def same_name(a: str, b: str) -> bool:
    """Case- and whitespace-insensitive sibling-name comparison."""
    return a.strip().casefold() == b.strip().casefold()


def _validate_fields(
    level: Level,
    name: str,
    description: str,
    business_owner: str,
    technical_owner: str | None,
) -> dict[str, str]:
    fields: dict[str, str] = {}
    if not (name or "").strip():
        fields["name"] = "Name is required"
    if not (description or "").strip():
        fields["description"] = "Description is required"
    if not (business_owner or "").strip():
        fields["businessOwner"] = "Business owner is required"
    if TECHNICAL_OWNER_REQUIRED[level] and not (technical_owner or "").strip():
        fields["technicalOwner"] = "Technical owner is required"
    return fields


def validate_create(level: Level, data: dict) -> dict:
    """Validate a create payload; returns a normalized field dict."""
    name = data.get("name", "")
    description = data.get("description", "")
    business_owner = data.get("businessOwner", "")
    technical_owner = data.get("technicalOwner")
    fields = _validate_fields(level, name, description, business_owner, technical_owner)
    if fields:
        raise ValidationError(fields)
    return {
        "name": name.strip(),
        "description": description.strip(),
        "businessOwner": business_owner.strip(),
        "technicalOwner": normalize_technical_owner(technical_owner),
    }


def validate_update(level: Level, merged: dict) -> dict:
    """Validate the merged (existing + patch) state of a node."""
    return validate_create(level, merged)

File: server/src/validation.ts
import { z } from "zod";
import { Level, TECHNICAL_OWNER_REQUIRED, CreateInput, UpdateInput } from "./types.js";

/** Thrown when a payload fails validation. Carries a field map for the API. */
export class ValidationError extends Error {
  status = 400;
  fields: Record<string, string>;
  constructor(fields: Record<string, string>) {
    super("Validation failed");
    this.name = "ValidationError";
    this.fields = fields;
  }
}

/** Thrown when a sibling with the same name already exists under a parent. */
export class ConflictError extends Error {
  status = 409;
  constructor(message: string) {
    super(message);
    this.name = "ConflictError";
  }
}

/** Thrown when an entity cannot be found. */
export class NotFoundError extends Error {
  status = 404;
  constructor(message: string) {
    super(message);
    this.name = "NotFoundError";
  }
}

const required = (label: string) =>
  z
    .string({ required_error: `${label} is required`, invalid_type_error: `${label} is required` })
    .trim()
    .min(1, `${label} is required`);

function buildCreateSchema(level: Level) {
  const base = {
    name: required("Name"),
    description: required("Description"),
    businessOwner: required("Business Owner"),
  };
  if (TECHNICAL_OWNER_REQUIRED[level]) {
    return z.object({ ...base, technicalOwner: required("Technical Owner") });
  }
  // Technical owner optional: accept empty string / null / undefined.
  return z.object({
    ...base,
    technicalOwner: z.string().trim().optional().nullable(),
  });
}

function zodToFields(err: z.ZodError): Record<string, string> {
  const out: Record<string, string> = {};
  for (const issue of err.issues) {
    const key = String(issue.path[0] ?? "_");
    if (!out[key]) out[key] = issue.message;
  }
  return out;
}

/** Validate and normalize a create payload for the given level. */
export function validateCreate(level: Level, input: unknown): Required<CreateInput> {
  const result = buildCreateSchema(level).safeParse(input);
  if (!result.success) throw new ValidationError(zodToFields(result.error));
  const data = result.data as CreateInput;
  return {
    name: data.name,
    description: data.description,
    businessOwner: data.businessOwner,
    technicalOwner: normalizeTechnicalOwner(data.technicalOwner),
  } as Required<CreateInput>;
}

/**
 * Validate an update payload. Only provided fields are checked, but the
 * resulting record must still satisfy required-field rules, so callers pass
 * the merged (existing + patch) record for a full re-validation.
 */
export function validateUpdate(level: Level, merged: CreateInput): Required<CreateInput> {
  return validateCreate(level, merged);
}

export function normalizeTechnicalOwner(value: string | null | undefined): string | null {
  if (value == null) return null;
  const trimmed = value.trim();
  return trimmed.length ? trimmed : null;
}

/** Case-insensitive, whitespace-insensitive comparison for sibling names. */
export function sameName(a: string, b: string): boolean {
  return a.trim().toLowerCase() === b.trim().toLowerCase();
}
